volume-group-operator/controllers/volumegroup_controller.go

```diff
@@ -53,14 +61,184 @@ type VolumeGroupReconciler struct {
 //+kubebuilder:rbac:groups="",resources=persistentvolumeclaims/finalizers,verbs=update
 
 func (r *VolumeGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := r.Log.WithValues("Request.Name", req.Name, "Request.Namespace", req.Namespace)
+
+	instance := &volumegroupv1.VolumeGroup{}
+	err := r.Client.Get(context.TODO(), req.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+
+			logger.Info("VolumeGroup resource not found")
+
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	vgcObj, err := r.Utils.GetVolumeGroupClass(logger, *instance.Spec.VolumeGroupClassName)
+	if err != nil {
+		uErr := r.updateVolumeGroupStatusError(instance, logger, err.Error())
+		if uErr != nil {
+			logger.Error(uErr, "failed to update volumeGroup status", "VGName", instance.Name)
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	if r.DriverConfig.DriverName != vgcObj.Driver {
+		return ctrl.Result{}, nil
+	}
+
+	err = utils.ValidatePrefixedParameters(vgcObj.Parameters)
+	if err != nil {
+		logger.Error(err, "failed to validate parameters of volumegroupClass", "VGClassName", vgcObj.Name)
+		uErr := r.updateVolumeGroupStatusError(instance, logger, err.Error())
+		if uErr != nil {
+			logger.Error(uErr, "failed to update volumeGroup status", "VGName", instance.Name)
+		}
+
+		return ctrl.Result{}, err
+	}
+	parameters := utils.FilterPrefixedParameters(utils.VolumeGroupParameterPrefix, vgcObj.Parameters)
+
+	secretName := vgcObj.Parameters[utils.PrefixedVolumeGroupSecretNameKey]
+	secretNamespace := vgcObj.Parameters[utils.PrefixedVolumeGroupSecretNamespaceKey]
```

Comment:
what will happen when the parameters in the volumeGroupClass won't have those keys? it will fail?

Reply:
no.. its only check for params with those keys from the first place

Reply:
where do we validate that it checks for params with those keys before?

Reply:
before what? I don't understand the question, those are special prefixes to the sidecar, we are checking that they are not empty.. every other parameter is optional and we don't care about it.
it is not suggested that the secret is mandatory, just that the parameter that we managing is valid.

Reply:
never mind my bad